File: src/flowa/artifact.py
"""Shared schema primitives for aggregation prompt sets.

`CategoryResult` is the citation-grounded base each deployment derives from
to add its own classification field(s) — e.g. generic ACMG sets add
`classification` + `classification_rationale`; private deployments may add
their own classification dimensions. Each deployment also defines its own
`AggregationResult` with `results: list[CategoryResult]` pointing at its
own `CategoryResult` subclass (Pydantic doesn't covariantly substitute the
parent's parameterised generic).

This module is the Python analog of `@flowajs/chat-service`'s
`artifactFields` / `ArtifactSchema` export: it exists so prompt-set schemas
share their citation-grounded structure rather than duplicating it.

The schema-side descriptions here are deliberately structural — they
describe what each field IS. Behavioural guidance (how to populate
`description`, ranking conventions, claim-emission criteria) belongs in the
aggregation prompt template, not the schema.
"""

from pydantic import BaseModel, Field


class AggregateCitation(BaseModel):
    """A citation quoting a specific passage from a source paper."""

    quote: str = Field(
        description='A short, distinctive verbatim quote from the paper text that identifies this evidence.'
    )


class Claim(BaseModel):
    """A factual statement supporting a synthesis, sourced from one paper."""

    paper_id: str = Field(description='AuthorYear identifier of the source paper; must appear in papers[].')
    text: str = Field(
        description='The factual statement as the curator reads it in triage. '
        'May synthesise across citations from the same paper.'
    )
    citations: list[AggregateCitation] = Field(
        description='One or more supporting quotes from paper_id.',
        min_length=1,
    )


class RankedPaper(BaseModel):
    """A paper in the ranked papers list. List position encodes importance."""

    paper_id: str = Field(description='AuthorYear identifier; must match the key used in paper_id_mapping.')
    rank_rationale: str = Field(description='One sentence explaining why this paper sits at this rank.')


class CategoryResult(BaseModel):
    """Citation-grounded base for an aggregation result.

    Deployments subclass to add classification-specific fields. The five
    fields here are the contract that `flowa.aggregate` reads at validation
    time.
    """

    category: str = Field(description='Assessment-category identifier.')
    description: str = Field(description='Short human-readable summary of the synthesis.')
    notes: str = Field(
        description='Long-form synthesis in Markdown. Uses inline citation links: '
        '[text](#cite:paper_id "verbatim quote") to reference specific paper evidence locations.'
    )
    papers: list[RankedPaper] = Field(
        description='Source documents contributing to the synthesis, ordered by importance. '
        'List position is the rank; paper_id values must be unique.'
    )
    claims: list[Claim] = Field(
        description='Factual claims supporting the synthesis, grouped by paper_id in the same order as papers[].'
    )

File: packages/chat-service/src/artifact.ts
// Citation-grounded artifact core. chat-service hardcodes a small fixed
// set of fields (`category`, `description`, `notes`, `papers`, `claims`)
// because those are what the chat-tool design — `view`, `search`,
// `str_replace`, `insert`, `write` over a citation-grounded synthesis —
// actually depends on. Deployments extend by calling
// `ArtifactSchema.extend({...})` with their additional fields; they do
// not rename the core fields.
//
// Field names mirror flowa's `prompts/generic/aggregation_schema.py`.

import { z } from "zod";

const CitationBbox = z.object({
  page: z.number().describe("Page number in the source document."),
  top: z.number().describe("Top edge of the highlight rectangle."),
  left: z.number().describe("Left edge of the highlight rectangle."),
  bottom: z.number().describe("Bottom edge of the highlight rectangle."),
  right: z.number().describe("Right edge of the highlight rectangle."),
});

const CitationMarkdownAnchor = z.object({
  start: z
    .number()
    .describe("Start offset (Unicode code point) of the quote in markdown.md."),
  end: z
    .number()
    .describe(
      "End offset (exclusive, Unicode code point) of the quote in markdown.md.",
    ),
});

const CitationLocation = z.object({
  bboxes: z
    .array(CitationBbox)
    .describe(
      "PDF highlight rectangles for the quote; empty when it wasn't aligned in the source PDF.",
    ),
  markdown_anchor: CitationMarkdownAnchor.nullable().describe(
    "Half-open [start, end) code-point range of the quote in markdown.md; null when it wasn't located there.",
  ),
});

const ArtifactCitation = z.object({
  quote: z
    .string()
    .describe(
      "Verbatim passage from the source document. Must be copied exactly — enough context to validate the claim.",
    ),
  location: CitationLocation.nullish().describe(
    "Pre-computed resolution of the quote: where it lives in the source PDF (bboxes) and the assembled markdown.md (anchor). Present on pipeline-generated citations (possibly with empty bboxes / null anchor); absent on citations added during editing.",
  ),
});

const ArtifactClaim = z.object({
  paper_id: z
    .string()
    .describe(
      "Identifier of the source document. Must appear in the papers[] list.",
    ),
  text: z
    .string()
    .describe(
      "The factual statement supported by the source document. One atomic fact per claim.",
    ),
  citations: z
    .array(ArtifactCitation)
    .min(1)
    .describe(
      "One or more supporting verbatim quotes from the document identified by paper_id.",
    ),
});

const RankedPaper = z.object({
  paper_id: z.string().describe("Identifier of the source document."),
  rank_rationale: z
    .string()
    .describe(
      "One sentence explaining why this document sits at this rank in the synthesis.",
    ),
});

/**
 * Citation-grounded fields shared by every chat-service artifact. Deployments
 * extend the schema by calling `ArtifactSchema.extend({...})` with their
 * additional fields:
 *
 * ```ts
 * import { ArtifactSchema } from "@flowajs/chat-service";
 *
 * export const MyArtifactSchema = ArtifactSchema.extend({
 *   classification: z.string(),
 *   classification_rationale: z.string(),
 * });
 * ```
 *
 * `.extend(...)` preserves the shape's TypeScript type, so the resulting
 * schema stays assignable to `z.ZodType<Artifact>` at the
 * `createApp({ schema })` call site. Spreading `...artifactFields` into a
 * fresh `z.object({...})` works at runtime but widens the inferred shape
 * to `Record<string, any>`, which fails the static type. `artifactFields`
 * is exported anyway for runtime inspection / reflection.
 *
 * Deployments must keep these fields with these names; chat-service reads
 * them directly. Refining their types (e.g. constraining `category` to a
 * Zod enum) is fine; renaming is not.
 */
export const artifactFields = {
  category: z
    .string()
    .describe(
      "Selector identifying which result this artifact represents within its aggregate. Mirrors aggregation.results[].category in flowa's prompts/generic/aggregation_schema.py.",
    ),
  description: z
    .string()
    .describe("Short human-readable summary of the synthesis."),
  notes: z
    .string()
    .describe(
      'Long-form synthesis in Markdown. Uses inline citation links: [text](#cite:paper_id "verbatim quote").',
    ),
  papers: z
    .array(RankedPaper)
    .describe(
      "Source documents contributing to the synthesis, ordered by contribution. List position is the rank. paper_id values must be unique.",
    ),
  claims: z
    .array(ArtifactClaim)
    .describe(
      "Factual claims supporting the synthesis. Grouped by paper_id in the same order as papers[].",
    ),
} as const;

/**
 * The minimal Zod schema chat-service can operate on. Any plugin's schema is
 * a superset: it spreads `artifactFields` and adds deployment fields.
 */
export const ArtifactSchema = z.object(artifactFields);

/**
 * The citation-grounded artifact type. Deployments may name their own
 * extended type; this is the base chat-service operates on.
 */
export type Artifact = z.infer<typeof ArtifactSchema>;

/**
 * Render a Zod schema as a JSON Schema string for inclusion in the system
 * prompt. The descriptions on each field serve as documentation.
 */
export function schemaForPrompt(schema: z.ZodType): string {
  return JSON.stringify(z.toJSONSchema(schema), null, 2);
}
